niri-ipc/src/lib.rs

```diff
@@ -1047,6 +1047,90 @@ pub enum OutputAction {
     },
 }
 
+macro_rules! ensure {
+    ($cond:expr, $fmt:literal $($arg:tt)* ) => {
+        if !$cond {
+            return Err(format!($fmt $($arg)*).into());
+        }
+    };
+}
+
+impl OutputAction {
+    /// Validates extra constraints that were difficult to represent with clap.
+    pub fn validate(&self) -> Result<(), String> {
+        match self {
+            OutputAction::Modeline {
+                hdisplay,
+                hsync_start,
+                hsync_end,
+                htotal,
+                vdisplay,
+                vsync_start,
+                vsync_end,
+                vtotal,
+                ..
+            } => {
+                ensure!(
+                    hdisplay < hsync_start,
+                    "hdisplay {} must be < hsync_start {}",
+                    hdisplay,
+                    hsync_start
+                );
+                ensure!(
+                    hsync_start < hsync_end,
+                    "hsync_start {} must be < hsync_end {}",
+                    hsync_start,
+                    hsync_end,
+                );
+                ensure!(
+                    hsync_end < htotal,
+                    "hsync_end {} must be < htotal {}",
+                    hsync_end,
+                    htotal,
+                );
+                ensure!(
+                    0u16 < *htotal,
+                    "htotal {} > 0",
+                    htotal,
+                );
+                ensure!(
+                    vdisplay < vsync_start,
+                    "vdisplay {} must be < vsync_start {}",
+                    vdisplay,
+                    vsync_start,
+                );
+                ensure!(
+                    vsync_start < vsync_end,
+                    "vsync_start {} must be < vsync_end {}",
+                    vsync_start,
+                    vsync_end,
+                );
+                ensure!(
+                    vsync_end < vtotal,
+                    "vsync_end {} must be < vtotal {}",
+                    vsync_end,
+                    vtotal,
+                );
+                ensure!(
+                    0u16 < *vtotal,
+                    "vtotal {} > 0",
+                    vtotal,
+                );
+                Ok(())
+            }
+            OutputAction::CustomMode {
+                mode: ConfiguredMode { refresh, .. },
+            } => {
+                if refresh.is_none() {
```

Comment:
Should also check for `== Some(0)`?

Reply:
Is it okay to require > 0 inside the `FromStr for ConfiguredMode` impl ?

Reply:
Let's not (unless custom=true), that would be a breaking change. I believe 0 currently means pick any refresh rate